Connection Node Failover

lipmas opened the pull request:
## Description of the change

Daemon changes:
- The `DaemonConnectionAuthProvider` is going to be used by new version daemons to get a connection node from the orchestrator. The provider will also validate the `DaemonConnectionAuthToken` and refresh it when it needs to by using the `api/v2/connections/{id}/auth-details` endpoint in the bastion.
- The `DaemonConnectionNodeAuthDetails` stores the `ConnectionNodeAuthToken` and the `ConnectionNodeUrl` that the orchestrator returns. This auth token is used for authorization when the daemon connects to the connection node.
- When a daemon websocket dies, the daemon will go into a backoff loop of trying to get a new connection node from the orchestrator and connect to it. (The same process as when it successfully connects the first time)

Agent changes:

- Augments the logic inside the control channel processing of the `openWebsocket` message to check if there is an existing dataconnection for the connection ID and if one already exists calls the new `UpdateConnectionNodeUrl` method. This will update the connection node url used in the underlying websocket connection and then restart the websocket connection.

-  When agent is restarting we now close all dataconnections with a specific error `ControlChannelClosingError` which is handled by only closing any active datachannels within the connection but otherwise leaving the connection open. This allows the daemon to use the new reconnect flow to resume the connection when the agent comes back online. This allows kube/db/web connections to work across an agent restart without the client needing to open a new connection. 

Zli PR: https://github.com/bastionzero/zli/pull/708
Backend PR: https://github.com/bastionzero/webshell-backend/pull/1608

## Testing

Testing instructions can be found in the backend PR linked above.

**backend branch:** feat/connection-node-failover
**zli branch:** feat/connection-node-failover
**charts branch:** 

#### Ready to run system tests?

- [x] Yes

## Relevant release note information

Release Notes: Connection Node Failover daemon/agent reconnect flow

## Related JIRA tickets

Relates to JIRA: CWC-2680

## Have you considered the security impacts?

Does this PR have any security impact?

- [ ] Yes
- [ ] No

If yes, please explain:

lipmas commented on bctl/agent/main.go:
```diff
@@ -463,7 +463,7 @@ func NewKubeAgent(
 
 func getAgentVersion() string {
-	return "$AGENT_VERSION"
+	return "8.0.0"
```
Reminder to revert this change before merge.

lipmas commented on bctl/daemon/daemon.go:
```diff
@@ -32,7 +32,7 @@ import (
 
 const (
-	daemonVersion  = "$DAEMON_VERSION"
+	daemonVersion  = "8.0.0"
```
Reminder to revert this change before merge.

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,208 @@
+	return token, nil
+}
+
+func (d *DaemonConnectionAuthProvider) getAuthDetailsFromBastion(ctx context.Context) (BastionConnectionAuthDetails, error) {
```
I believe thanos introduced a new [bastion](https://github.com/bastionzero/bzero/blob/develop/bctl/daemon/bastion/bastion.go) entity for making these kinds of queries so that we can consolidate authentication to bastion etc.. Would it make sense to move this there?

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,208 @@
+
+func (d *DaemonConnectionAuthProvider) getAuthDetailsFromBastion(ctx context.Context) (BastionConnectionAuthDetails, error) {
+	// Load id_token dynamically from zli config in case it has changed (login or refresh)
+	zliConfig, err := zliconfig.New(d.zliConfigPath, d.logger)
```
this feels like a variable that should be passed in at instantiation for easy mocking

johncmerfeld reviewed (approved):
This is great. I performed the test as described in the backend PR and it worked out.

I also tested something that has been a major risk factor for us: the case where a daemon is hard-killed and can't send a `CloseDataChannel` message. If this was done while the agent had messages to send, it would start spamming the backend and build up messages in its queue, eventually leading to a crash.

YOU FIXED THIS. Here's what I did to verify:
  - `zli connect` to get a shell
  - Run `yes` in the shell
  - Find the daemon process and kill it with `kill -9` to prevent any graceful shutdown
  - Observe the agent's memory with `top -p `pgrep agent | tr "\\n" "," | sed 's/,$//'``

Since the `yes` command is still running, the agent is still reading messages and trying to send them. But now at least it is dropping them because the daemon is disconnected, so this is no longer a dangerous, unrecoverable issue. Super excited about this. I wish we could automate this test

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,208 @@
+	return &responseDecoded, nil
+}
+
+func (d *DaemonConnectionAuthProvider) refreshToken(ctx context.Context) error {
```
I see how this makes sense, but I think we should be creating a really strong dilineation between the auth token and the auth details (which I think is done really well in other places). IMO, I think we could call this function `refreshAuth`?

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+func (d *DaemonConnectionAuthProvider) verifyToken(idToken string, ctx context.Context) (*oidc.IDToken, error) {
```
in line with my comment above, but I think this would be clearer as `verifyIdToken`

lipmas commented on bctl/agent/controlchannel/dataconnection/connection.go:
```diff
@@ -213,15 +264,17 @@ func (d *DataConnection) receive() {
 					// message so that this timeout is configurable from the
 					// backend per connection
 					d.idleTimeout = dcMessage.IdleTimeout.Duration
+					d.daemonConnectedChan <- struct{}{}
```
Well i tried to ensure that the tmb go function in `New()` is now the only place calling `waitForDaemonConnected()` and only place where we read from the channel. So I think there shouldnt be a race anymore since all the reads from the channel happen within this single go routine.

lipmas commented on bctl/daemon/daemon.go:
```diff
@@ -98,6 +98,7 @@ func createLogger(debug bool) (*bzlogger.Logger, error) {
 
 	logger, err := bzlogger.New(options)
 	logger.AddDaemonVersion(daemonVersion)
+	logger = logger.GetConnectionLogger(config[CONNECTION_ID].Value)
```
So it is a required daemon argument [here](https://github.com/bastionzero/bzero/blob/82588ebef3d4df2e0cec59fc8c66316f677aa4b8/bctl/daemon/config.go#L55) so i think it should always be set otherwise the daemon would exit. I was trying to setup the logger context to include the actual connection ID that we can correlate with backend logs. Putting this as early in the daemon flow as possible made the most sense to me.

johncmerfeld replied:
Makes sense! I blanked and forgot we checked for required args

lgmugnier reviewed (commented):
I did some refactoring to help me dig into and understand the code better and I've shared those results separately but such a refactor is absolutely not necessary to help clarify. My comments here focus on the specifics of what I found helpful plus some small things I found when digging into the code.

lipmas commented on bctl/daemon/servers/dataconnection/connection.go:
```diff
@@ -211,6 +207,10 @@ func New(
 				// processes the new AgentConnected message.
 				conn.waitForAgentReady()
+			case <-conn.agentDisconnectedChan:
+				conn.logger.Infof("Agent disconnected waiting for agent to reconnect for %s before triggering reconnect flow", agentConnectedTimeout)
```
[b70d370](https://github.com/bastionzero/bzero/pull/251/commits/b70d37084d7c41ebfa794b8503aeac0302b12f71)

lipmas commented:
> I did some refactoring to help me dig into and understand the code better and I've shared those results separately but such a refactor is absolutely not necessary to help clarify. My comments here focus on the specifics of what I found helpful plus some small things I found when digging into the code.

Ok so i took a pass at refactoring the code to hopefully make the methods/interfaces in the `daemonconnauth` package more clear. I also pulled some of your ideas from your experimental branch and moved the `DaemonConnectionAuthToken` refresh logic that interacts with bastion into the bastion package and refactored bastion code to be responsible for refreshing the IdP auth tokens when they expire.
 
[d4cd366](https://github.com/bastionzero/bzero/pull/251/commits/d4cd3664be166884f31616785bcf644a96167c28)

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,148 @@
+	GetDaemonConnectionAuthToken(ctx context.Context, connectionId string) (authToken string, connectionServiceUrl string, err error)
+}
+
+type DaemonConnectionAuth struct {
```
This is how I see it: you have 3 main functions in the corresponding struct:
`GetConnectionNodeAuthToken`
`refreshConnectionNodeAuthToken`
`verifyConnectionNodeAuthToken`

To me it sounds like you have an object called `ConnectionNodeAuthToken` (even though I know the struct contains some extra info) with three functions: `Get`, `refresh`, and `verify`. I personally think that makes the organization more concise and object-oriented and consistent with other token implementations in this repo. I also think it makes sense to keep the interface `DaemonConnectionAuth` with a single function `Get` since the way it is now feels a bit redundant. As long as the `daemonconnauth` package is clear about what token it's working with, it doesn't need to propagate all the key terms to the rest of the code.

lgmugnier replied:
oh you're right. ugh yeah I think this is just a consequence of having long token names (which I'm not in favor of changing). I think with the comments, it should help.

lgmugnier commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,148 @@
+	verifier := oidcProvider.Verifier(config)
+	token, err := verifier.Verify(ctx, idToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify id token with provider: %w", err)
```
we should mention that we're verifying the daemon connection auth token here

lipmas replied:
[e34b548](https://github.com/bastionzero/bzero/pull/251/commits/e34b5488ce7f5804d492217a302512252a5c7284)

lgmugnier reviewed (commented):
Wooo! I love what you've done with the place. I really appreciate the clarification and consistency between the different tokens and all the work you put into the bastion!

lipmas commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,148 @@
+	GetConnectionNodeAuthToken(ctx context.Context) (connectionNodeUrl string, connectionNodeAuthToken string, err error, retry bool)
+}
+
+type IBastionClient interface {
```
[0f46f41](https://github.com/bastionzero/bzero/pull/251/commits/0f46f41f3bf5d57e624e516606ad9e2446a2fe4e)

lipmas commented on bctl/daemon/bastion/bastion.go:
```diff
@@ -5,43 +5,75 @@ import (
 
-type ApiClient interface {
-	RefreshTokens(ctx context.Context, contentType string, idToken string, accessToken string) (sessionId string, sessionToken string, err error)
+type IBastionConfig interface {
```
[0f46f41](https://github.com/bastionzero/bzero/pull/251/commits/0f46f41f3bf5d57e624e516606ad9e2446a2fe4e)

lipmas commented on bctl/daemon/bastion/bastion.go:
```diff
@@ -84,3 +116,47 @@ func (b *Bastion) RefreshTokens(ctx context.Context, idToken string, accessToken
 	return nil, "", "", fmt.Errorf("failed to retrieve valid sessionId/token")
 }
+
+func (b *Bastion) GetDaemonConnectionAuthToken(ctx context.Context, connectionId string) (connectionServiceUrl string, authToken string, err error) {
```
[3e50a66](https://github.com/bastionzero/bzero/pull/251/commits/3e50a666b272074319c99c1b9325a7b64b616e0d)

lipmas commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,148 @@
+	}
+}
+
+func (d *DaemonConnectionAuth) GetConnectionNodeAuthToken(ctx context.Context) (connectionNodeUrl string, connectionNodeAuthToken string, err error, retry bool) {
```
[3e50a66](https://github.com/bastionzero/bzero/pull/251/commits/3e50a666b272074319c99c1b9325a7b64b616e0d)

lipmas commented on bctl/daemon/bastion/daemonconnauth/daemonconnauth.go:
```diff
@@ -0,0 +1,148 @@
+)
+
+type IDaemonConnectionAuth interface {
+	GetConnectionNodeAuthToken(ctx context.Context) (connectionNodeUrl string, connectionNodeAuthToken string, err error, retry bool)
```
[0614781](https://github.com/bastionzero/bzero/pull/251/commits/061478119711e9c2a68c326ae064c61daf1fd377)

lgmugnier reviewed (approved):
Thanks for bearing with all of my questions and concerns on this! I'm super happy where we've ended up and it's probably as clear as we could make it for right now which is important considering the system complexity of this code.
